--- a/portal/data/meson.build
+++ b/portal/data/meson.build
@@ -9,15 +9,6 @@ if compile_schemas.found()
   )
 endif
 
-systemduserunitdir = get_option('systemduserunitdir')
-if systemduserunitdir == ''
-  systemd = dependency('systemd', version: '>= 242')
-  systemduserunitdir = systemd.get_pkgconfig_variable(
-    'systemduserunitdir',
-    define_variable: ['prefix', get_option('prefix')]
-  )
-endif
-
 dbus_service_dir = get_option('dbus_service_dir')
 if dbus_service_dir == ''
   dbus_service_dir = datadir / 'dbus-1' / 'services'
@@ -26,14 +17,6 @@ endif
 portal_conf_data = configuration_data()
 portal_conf_data.set('libexecdir', libexec_dir)
 
-configure_file(
-  input: 'xdg-desktop-portal-impatience.service.in',
-  output: '@BASENAME@',
-  configuration: portal_conf_data,
-  install: true,
-  install_dir: systemduserunitdir
-)
-
 install_data(
   'impatience.portal',
   install_dir: join_paths(get_option('prefix'), get_option('datadir'), 'xdg-desktop-portal', 'portals')
